### PR TITLE
SDPA support user-provided mask when sequence lengths not chunk-aligned

### LINKED ISSUE #32502: SDPA fails with sequence lengths not divisible by 32
### Describe the bug

The `scaled_dot_product_attention` operation fails when query or key/value sequence lengths are not divisible by 32 and mask is present.

### Steps to reproduce the issue

Run the following test script:

```py
import pytest
import torch
import ttnn

@pytest.mark.parametrize(
    "q_shape,k_shape,v_shape,mask_shape",
    [
        # Query seq_len not divisible by 32
        ((1, 8, 63, 64), (1, 8, 64, 64), (1, 8, 64, 64), (1, 1, 63, 64)),
        # Key/Value seq_len not divisible by 32
        ((1, 8, 64, 64), (1, 8, 50, 64), (1, 8, 50, 64), (1, 1, 64, 50)),
        # Both query and key/value seq_len not divisible by 32
        ((1, 4, 100, 64), (1, 4, 77, 64), (1, 4, 77, 64), (1, 1, 100, 77)),
    ],
)
def test_sdpa_non_divisible_seqlen(q_shape, k_shape, v_shape, mask_shape):
    """Test SDPA with shapes that have sequence lengths not divisible by 32."""
    
    torch.manual_seed(42)
    
    Q = torch.randn(q_shape)
    K = torch.randn(k_shape)
    V = torch.randn(v_shape)
    attention_mask = torch.zeros(mask_shape)
    
    device = ttnn.open_device(device_id=0)
    
    try:
        tt_Q = ttnn.from_torch(Q, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
        tt_K = ttnn.from_torch(K, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
        tt_V = ttnn.from_torch(V, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
        tt_attention_mask = ttnn.from_torch(attention_mask, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
        
        tt_output = ttnn.transformer.scaled_dot_product_attention(
            tt_Q,
            tt_K,
            tt_V,
            is_causal=False,
            attn_mask=tt_attention_mask,
        )
        
        output = ttnn.to_torch(tt_output)
            
    finally:
        ttnn.close_device(device)
```

### Expected behavior

Support padding the sequence dim to be divisible by 32 and then slice the output back to the original dimensions.

### LINKED ISSUE #36097: Support attn_mask when Q/K have non-chunk-aligned sequence lengths in SDPA
### Is your feature request related to a problem? Please describe

The current `ttnn.transformer.scaled_dot_product_attention` op allows Q/K tensors with any sequence length when no mask is provided (internal padding is handled automatically). However, when an attn_mask is provided, Q and K sequence lengths must be divisible by q_chunk_size and k_chunk_size respectively.

For TT-Boltz, we have sequences with lengths that are not divisible by chunk size, and we need to provide an attention mask.

### Describe the solution you'd like

Allow providing attn_mask for Q/K tensors with any sequence length. The SDPA kernel should:
- Accept a mask matching the logical (unpadded) Q/K dimensions
- Read mask tiles for valid positions and fill padded K positions with -inf
- Work with all supported mask data types (BFLOAT16, BFLOAT8_B, BFLOAT4_B)

### Describe alternatives you've considered

Padding Q/K/mask to chunk-aligned lengths before SDPA - too much overhead.

### Additional context

I'll open a pull request today. 

CC: @cglagovichTT @sosborne-TT

### PR DESCRIPTION
**Pull request by moritztng**
### Ticket
Closes #36097 

### Problem description
The current `ttnn.transformer.scaled_dot_product_attention` op allows Q/K tensors with any sequence length when no mask is provided (internal padding is handled automatically). However, when an attn_mask is provided, Q and K sequence lengths must be divisible by q_chunk_size and k_chunk_size respectively.

### What's changed

When an `attn_mask` is provided with non-chunk-aligned Q/K sequences, the reader kernel now:
1. Reads from the user's unpadded mask using `valid_Sqt/valid_Skt` dimensions
2. Fills padded Q and K tile positions with negative infinity

Key changes:
- **`reader_interleaved.cpp`**: Mask indexing uses unpadded dimensions; tiles in padded regions are filled with `-inf` via `fill_neginf_tile`
- **`dataflow_common.hpp`**: Generic `fill_neginf_tile` function supporting bfp4, bfp8, and bfloat16 formats
- **`writer_interleaved.cpp`**: Padding mask generation now only runs when no user mask is provided
- **`sdpa_device_operation.cpp`**: Removed chunk-divisibility validation when mask is provided
- **`sdpa_program_factory.cpp`**: `use_padded_mask` now detects both Q and K padding

This enables non-causal SDPA with user-provided masks on any sequence length, matching the behavior of the no-mask case.

### Checklist

- [x] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml?query=branch:{{branch_name}})
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:{{branch_name}})
- [x] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:{{branch_name}})
- [ ] New/Existing tests provide coverage for changes

#### Model tests

If your changes cover model-related code, you should run tests corresponding to affected models and platforms (Single card, T3K, Galaxy). "Choose your pipeline" workflows facilitate running multiple kinds of tests in a single run. Each offers `models-mandatory` and `models-extended` presets.
The former includes a minimal set of tests, to be run always. The latter extends that with additional ones - use your best judgement in deciding which is the most appropriate for your PR.

- [ ] [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:{{branch_name}})
  - [ ] `models-mandatory` preset (runs: [Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) and [Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:{{branch_name}})
  - [x] `models-mandatory` preset (runs: [Unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:{{branch_name}})
  - [ ] `models-mandatory` preset (runs: [Quick tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml))
  - [ ] `models-extended` preset (runs: the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/dataflow/dataflow_common.hpp [hunk 1]
```diff
@@ -187,31 +187,45 @@ void copy_tile(uint64_t noc_read_addr_base, uint32_t q_write_ptr_base, uint32_t
...
+}
 
-    // Fill the next 512 bytes (128 uint32_t values) with 0xCCCCCCCC for mantissas
-    constexpr uint32_t NEG_INF_MANT = 0xCCCCCCCC;
-    constexpr uint32_t mant_words = num_mantissas / sizeof(uint32_t);  // 128 words
-
-    for (uint32_t i = exp_words; i < exp_words + mant_words; i++) {
-        ptr[i] = NEG_INF_MANT;
-    }
+// Keep bfp4 version for backward compatibility with existing causal mask generation
```
Comment (sosborne-TT) [inline ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/dataflow/dataflow_common.hpp]: The old version can be deleted and anywhere using it updated to call your new version.

**Review by sosborne-TT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/dataflow/dataflow_common.hpp (hunk 1 above)
Comment (moritztng) [reply ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/dataflow/dataflow_common.hpp]: Done

**Review by cglagovichTT — APPROVED** [review]

**Review by gwangTT — COMMENTED** [review]
Please run t3k unit test CI pipeline (to make sure qwen25_vl passes at least -- test_windowed_sdpa.py is run in there): https://github.com/tenstorrent/tt-metal/actions/runs/21396670511/job/61597477242

**Comment by moritztng** [comment]
> Please run t3k unit test CI pipeline (to make sure qwen25_vl passes at least -- test_windowed_sdpa.py is run in there): https://github.com/tenstorrent/tt-metal/actions/runs/21396670511/job/61597477242

I already ran t3k models-mandatory which includes the unit tests: https://github.com/tenstorrent/tt-metal/actions/runs/21336182321

The tests that failed also failed in main.

**Review by gwangTT — APPROVED** [review]
